src/scheduler/fifo.c

```diff
@@ -917,28 +917,38 @@ main_sched_loop(status *policy, int sd, server_info *sinfo, schd_error **rerr)
 				tj = njob;
 
 			if (rc != SCHD_ERROR) {
-				if(run_update_resresv(policy, sd, sinfo, qinfo, tj, ns_arr, RURR_ADD_END_EVENT, err) > 0 ) {
-					rc = SUCCESS;
-					if (sinfo->has_soft_limit || qinfo->has_soft_limit)
-						sort_again = MUST_RESORT_JOBS;
-					else
-						sort_again = MAY_RESORT_JOBS;
-				} else {
-					/* if run_update_resresv() returns 0 and pbs_errno == PBSE_HOOKERROR,
-					 * then this job is required to be ignored in this scheduling cycle
-					 */
-					rc = err->error_code;
+				if (!(tj->job->window_enabled)) {
+					rc = JOB_WINDOW_NOT_STARTED;
```

Comment:
This is the wrong place for checking the time window.  Put the check for the time window in is_ok_to_run().

Reply:
hey @bhroam - thank you for looking into this.
If we check in is_ok_to_run() and return NULL, we fall into the else part and try preemption. We need another check there. Another reason is that we need to set rc = JOB_WINDOW_NOT_STARTED, only if the job did not run because it's time window has not started, not for some other reason, so I added this check just before we send the job to the server.

Reply:
The check still belongs in is_ok_to_run().  The error code still comes back in the err structure.  Another reason this has to be in is_ok_to_run() is because of calendaring.  We need to be able to know when the job can run via is_ok_to_run().  Otherwise we'll add it on the calendar in places it can't run since it is out of the time window.  If you are worried about preemption, there is a section of code there that checks a set of static reasons that we bail out of the preemption code.  For instance if do_not_span_psets is on and we need to span.  Preemption can't help there.  Just add a check for being out of your time window there.

Reply:
done. there is some code refactoring needed still...